File: src/services/auth.service.js
import bcrypt from 'bcrypt';
import jwt from 'jsonwebtoken';
import UserModel from '../models/user.model.js';
import VerificationTokenModel from '../models/verificationToken.model.js';
import { query } from '../config/database.js';
import emailService from './email.service.js';
import logger from '../utils/logger.js';

const JWT_SECRET = process.env.JWT_SECRET;
const JWT_EXPIRES_IN = process.env.JWT_EXPIRES_IN || '7d';
const BCRYPT_ROUNDS = 10;

class AuthService {
  /**
   * Sanitize user object for returning to client
   * @param {Object} user - User object
   * @returns {Object} Sanitized user
   */
  static sanitizeUser(user) {
    if (!user) return null;
    const { password_hash, github_access_token, ...sanitizedUser } = user;
    return sanitizedUser;
  }

  /**
   * Hash password
   * @param {string} password - Plain text password
   * @returns {Promise<string>} Hashed password
   */
  static async hashPassword(password) {
    const salt = await bcrypt.genSalt(BCRYPT_ROUNDS);
    return bcrypt.hash(password, salt);
  }

  /**
   * Compare password
   * @param {string} password - Plain text password
   * @param {string} hash - Hashed password
   * @returns {Promise<boolean>} True if match
   */
  static async comparePassword(password, hash) {
    if (!password || !hash) return false;
    return bcrypt.compare(password, hash);
  }

  /**
   * Generate JWT token
   * @param {Object} payload - Token payload
   * @returns {string} JWT token
   */
  static generateToken(payload) {
    return jwt.sign(payload, JWT_SECRET, { expiresIn: JWT_EXPIRES_IN });
  }

  /**
   * Verify JWT token
   * @param {string} token - JWT token
   * @returns {Object} Decoded token
   */
  static verifyToken(token) {
    try {
      return jwt.verify(token, JWT_SECRET);
    } catch (error) {
      throw new Error('Invalid or expired token');
    }
  }

  /**
   * Register new user with email/password and send verification email
   * @param {Object} userData - User registration data
   * @returns {Promise<Object>} User and token
   */
  static async registerWithVerification(userData) {
    const { email, password, full_name } = userData;

    // Check if user already exists
    const existingUser = await UserModel.findByEmail(email);
    if (existingUser) {
      const error = new Error('User with this email already exists');
      error.statusCode = 400;
      throw error;
    }

    // Hash password
    const password_hash = await this.hashPassword(password);

    // Create user with email auth provider
    const user = await UserModel.create({
      email,
      password_hash,
      full_name,
      auth_provider: 'email',
      email_verified: false,
    });

    // Generate verification token (24 hours)
    const verificationToken = await VerificationTokenModel.create(
      user.id,
      'email_verification',
      60 * 24
    );

    // Send welcome email with verification link
    try {
      await emailService.sendWelcomeEmail(this.sanitizeUser(user), verificationToken.token);
      logger.info(`Welcome email sent to user: ${user.id}`);
    } catch (error) {
      logger.error(`Failed to send welcome email to ${email} (User ID: ${user.id}):`, error);
      // We don't throw here to avoid failing registration if only email fails
    }

    // Generate JWT token
    const token = this.generateToken({
      userId: user.id,
      email: user.email,
    });

    logger.info(`New user registered and verification email sent: ${email}`);

    return {
      user: this.sanitizeUser(user),
      token,
    };
  }



  /**
   * Login user with email/password
   * @param {string} email - User email
   * @param {string} password - User password
   * @returns {Promise<Object>} User and token
   */
  static async login(email, password) {
    // Find user using method that returns password_hash
    const user = await UserModel.findByEmailWithPassword(email);
    if (!user) {
      const error = new Error('Invalid email or password');
      error.statusCode = 401;
      throw error;
    }

    // Check if user has password (not OAuth-only user)
    if (!user.password_hash) {
      const error = new Error('Please login using GitHub OAuth');
      error.statusCode = 401;
      throw error;
    }

    // Verify password
    const isPasswordValid = await this.comparePassword(password, user.password_hash);
    if (!isPasswordValid) {
      const error = new Error('Invalid email or password');
      error.statusCode = 401;
      throw error;
    }

    // Update last login
    await UserModel.updateLastLogin(user.id);

    // Generate token
    const token = this.generateToken({
      userId: user.id,
      email: user.email,
    });

    logger.info(`User logged in: ${email}`);

    return {
      user: this.sanitizeUser(user),
      token,
    };
  }

  /**
   * Request password reset
   * @param {string} email - User email
   * @returns {Promise<boolean>} Success indicator
   */
  static async requestPasswordReset(email) {
    const user = await UserModel.findByEmail(email);

    // To prevent email enumeration, we return true even if user not found
    if (!user) {
      logger.warn(`Password reset requested for non-existent email: ${email}`);
      return true;
    }

    // Only allow password reset for users with password_hash
    if (user.auth_provider !== 'email') {
      logger.warn(`Password reset requested for OAuth user: ${email}`);
      return true; // Still return true for security
    }

    // Generate reset token (15 mins)
    const resetToken = await VerificationTokenModel.create(
      user.id,
      'password_reset',
      15
    );

    // Send reset email
    try {
      await emailService.sendPasswordResetEmail(this.sanitizeUser(user), resetToken.token);
      logger.info(`Password reset email sent to user: ${user.id}`);
    } catch (error) {
      logger.error(`Failed to send password reset email to ${email}:`, error);
      throw new Error('Failed to send reset email. Please try again later.');
    }

    return true;
  }

  /**
   * Reset password using token
   * @param {string} token - Reset token
   * @param {string} newPassword - New password
   * @returns {Promise<boolean>} Success indicator
   */
  static async resetPassword(token, newPassword) {
    // Find valid token
    const verificationToken = await VerificationTokenModel.findByToken(token, 'password_reset');

    if (!verificationToken) {
      const error = new Error('Invalid or expired reset token');
      error.statusCode = 400;
      throw error;
    }

    if (new Date(verificationToken.expires_at) < new Date()) {
      const error = new Error('Reset token has expired');
      error.statusCode = 400;
      throw error;
    }

    // Hash new password
    const password_hash = await this.hashPassword(newPassword);

    // Update user password
    await UserModel.update(verificationToken.user_id, {
      password_hash
    });

    // Mark token as used
    await VerificationTokenModel.markAsUsed(verificationToken.id);

    logger.info(`Password reset completed for user ID: ${verificationToken.user_id}`);

    return true;
  }

  /**
   * Verify email address
   * @param {string} token - Verification token
   * @returns {Promise<boolean>} Success indicator
   */
  static async verifyEmail(token) {
    const verificationToken = await VerificationTokenModel.findByToken(token, 'email_verification');

    if (!verificationToken) {
      const error = new Error('Invalid or expired verification token');
      error.statusCode = 400;
      throw error;
    }

    if (new Date(verificationToken.expires_at) < new Date()) {
      const error = new Error('Verification token has expired');
      error.statusCode = 400;
      throw error;
    }

    // Update user email_verified status
    await UserModel.update(verificationToken.user_id, {
      email_verified: true
    });

    // Mark token as used
    await VerificationTokenModel.markAsUsed(verificationToken.id);

    logger.info(`Email verified for user ID: ${verificationToken.user_id}`);

    return true;
  }

  /**
   * Resend verification email
   * @param {string} email - User email
   * @returns {Promise<boolean>} Success indicator
   */
  static async resendVerification(userId) {
    const user = await UserModel.findById(userId);

    if (!user) {
      const error = new Error('User not found');
      error.statusCode = 404;
      throw error;
    }

    if (user.email_verified) {
      const error = new Error('Email is already verified');
      error.statusCode = 400;
      throw error;
    }

    // Check if there's a recent active token to prevent spam
    const existingTokens = await query(
      `SELECT COUNT(*) FROM verification_tokens
       WHERE user_id = $1 AND type = 'email_verification'
       AND used = false AND expires_at > NOW()
       AND created_at > NOW() - INTERVAL '15 minutes'`,
      [user.id]
    );

    if (parseInt(existingTokens.rows[0].count) > 0) {
      const error = new Error('Please wait 15 minutes before requesting another verification email');
      error.statusCode = 429;
      throw error;
    }

    // Generate new verification token (24 hours)
    const verificationToken = await VerificationTokenModel.create(
      user.id,
      'email_verification',
      60 * 24
    );

    // Send email
    try {
      await emailService.sendVerificationEmail(this.sanitizeUser(user), verificationToken.token);
      logger.info(`Resent verification email to user: ${user.id}`);
    } catch (error) {
      logger.error(`Failed to resend verification email to ${user.email}:`, error);
      throw new Error('Failed to send verification email. Please try again later.');
    }

    return true;
  }

  /**
   * Handle GitHub OAuth login/registration
   * @param {Object} githubProfile - GitHub user profile data
   * @returns {Promise<Object>} User and token
   */
  static async handleGithubAuth(githubProfile) {
    const { id: githubId, username, email, name, avatarUrl, accessToken } = githubProfile;

    // First try to find user by GitHub ID
    let user = await UserModel.findByGithubId(githubId);

    if (user) {
      // User exists, update token and last login
      await UserModel.update(user.id, {
        github_access_token: accessToken,
        github_username: username,
        avatar_url: user.avatar_url || avatarUrl,
        last_login: new Date()
      });
      user = await UserModel.findById(user.id);
      logger.info(`User logged in via GitHub: ${user.id}`);
    } else {
      // Check if user exists with this email
      if (email) {
        user = await UserModel.findByEmail(email);

        if (user) {
          // Link GitHub to existing account
          user = await UserModel.linkGithubAccount(user.id, githubProfile);
          logger.info(`GitHub linked to existing account: ${user.id}`);
        } else {
          // Create new user
          user = await UserModel.create({
            email: email || `${username}@users.noreply.github.com`,
            full_name: name || username,
            github_id: githubId,
            github_username: username,
            github_access_token: accessToken,
            avatar_url: avatarUrl,
            auth_provider: 'github'
          });
          logger.info(`New user created via GitHub: ${user.id}`);
        }
      } else {
        // Create new user without email (rare from GitHub if scopes are right)
        user = await UserModel.create({
          email: `${username}@users.noreply.github.com`,
          full_name: name || username,
          github_id: githubId,
          github_username: username,
          github_access_token: accessToken,
          avatar_url: avatarUrl,
          auth_provider: 'github'
        });
        logger.info(`New user created via GitHub (no email): ${user.id}`);
      }
    }

    // Generate token
    const token = this.generateToken({
      userId: user.id,
      email: user.email,
    });

    return {
      user: this.sanitizeUser(user),
      token,
    };
  }

  /**
   * Ensure JWT_SECRET is configured
   */
  static validateConfig() {
    if (!JWT_SECRET) {
      logger.error('CRITICAL: JWT_SECRET environment variable is not set!');
      throw new Error('Server configuration error: JWT_SECRET is missing');
    }
  }
}

// Ensure config is valid at startup
AuthService.validateConfig();

export default AuthService;

File: app/services/auth_service.py
"""Authentication service."""
import bcrypt
from datetime import datetime, timedelta
from flask_jwt_extended import create_access_token, create_refresh_token
from app.extensions import db
from app.models import User, VerificationToken
from app.utils.logger import get_logger
from app.services.email_service import EmailService

logger = get_logger(__name__)
BCRYPT_ROUNDS = 10


class AuthService:
    """Service for authentication operations."""

    @staticmethod
    def hash_password(password):
        """Hash password with bcrypt."""
        try:
            salt = bcrypt.gensalt(rounds=BCRYPT_ROUNDS)
            return bcrypt.hashpw(password.encode("utf-8"), salt).decode("utf-8")
        except Exception as e:
            logger.error(f"Failed to hash password: {str(e)}")
            raise

    @staticmethod
    def verify_password(plain_password, hashed_password):
        """Verify password against hash."""
        try:
            if not hashed_password:
                return False
            return bcrypt.checkpw(plain_password.encode("utf-8"), hashed_password.encode("utf-8"))
        except Exception as e:
            logger.error(f"Failed to verify password: {str(e)}")
            return False

    @staticmethod
    def generate_tokens(user_id, email):
        """Generate JWT access and refresh tokens."""
        try:
            access_token = create_access_token(identity=user_id)
            refresh_token = create_refresh_token(identity=user_id)
            return {"access_token": access_token, "refresh_token": refresh_token}
        except Exception as e:
            logger.error(f"Failed to generate tokens: {str(e)}")
            raise

    @staticmethod
    def register_user(email, password, full_name):
        """Register new user with email verification."""
        try:
            # Check if user exists
            existing_user = User.query.filter_by(email=email).first()
            if existing_user:
                raise ValueError("User with this email already exists")

            # Hash password
            password_hash = AuthService.hash_password(password)

            # Create user
            user = User(
                email=email,
                password_hash=password_hash,
                full_name=full_name,
                auth_provider="email",
                email_verified=False,
            )
            db.session.add(user)
            db.session.commit()

            logger.info(f"User registered: {email}")

            # Create verification token
            verification_token = AuthService.create_verification_token(user.id, "email_verification", 24)

            # Send verification email
            try:
                EmailService.send_welcome_email(user, verification_token.token)
            except Exception as e:
                logger.error(f"Failed to send welcome email to {email}: {str(e)}")
                # Don't fail registration if email fails

            # Generate JWT tokens
            tokens = AuthService.generate_tokens(user.id, user.email)

            return {
                "user": user.to_dict(),
                "tokens": tokens,
                "message": "Registration successful. Please check your email to verify your account.",
            }

        except Exception as e:
            db.session.rollback()
            logger.error(f"Failed to register user: {str(e)}")
            raise

    @staticmethod
    def login_user(email, password):
        """Login user with email/password."""
        try:
            # Find user
            user = User.query.filter_by(email=email).first()
            if not user:
                raise ValueError("Invalid email or password")

            # Check password
            if not user.password_hash or not AuthService.verify_password(password, user.password_hash):
                raise ValueError("Invalid email or password")

            # Update last login
            user.last_login = datetime.utcnow()
            db.session.commit()

            # Generate tokens
            tokens = AuthService.generate_tokens(user.id, user.email)

            logger.info(f"User logged in: {email}")

            return {
                "user": user.to_dict(),
                "tokens": tokens,
            }

        except Exception as e:
            logger.error(f"Login failed for {email}: {str(e)}")
            raise

    @staticmethod
    def verify_email(token_string):
        """Verify user email with token."""
        try:
            # Find token
            token = VerificationToken.query.filter_by(
                token=token_string, token_type="email_verification"
            ).first()

            if not token:
                raise ValueError("Invalid verification token")

            if token.is_expired():
                raise ValueError("Verification token has expired")

            if token.is_used:
                raise ValueError("Verification token has already been used")

            # Update user
            user = User.query.get(token.user_id)
            if not user:
                raise ValueError("User not found")

            user.email_verified = True
            token.is_used = True
            token.used_at = datetime.utcnow()
            db.session.commit()

            logger.info(f"Email verified for user: {user.email}")

            return {"user": user.to_dict(), "message": "Email verified successfully"}

        except Exception as e:
            db.session.rollback()
            logger.error(f"Failed to verify email: {str(e)}")
            raise

    @staticmethod
    def request_password_reset(email):
        """Request password reset token."""
        try:
            # Find user
            user = User.query.filter_by(email=email).first()
            if not user:
                raise ValueError("User not found")

            # Create reset token (15 minutes)
            reset_token = AuthService.create_verification_token(user.id, "password_reset", 15)

            # Send reset email
            try:
                EmailService.send_password_reset_email(user, reset_token.token)
            except Exception as e:
                logger.error(f"Failed to send password reset email to {email}: {str(e)}")
                raise

            logger.info(f"Password reset requested for: {email}")

            return {"message": "Password reset email sent"}

        except Exception as e:
            logger.error(f"Failed to request password reset: {str(e)}")
            raise

    @staticmethod
    def reset_password(token_string, new_password):
        """Reset password with token."""
        try:
            # Find token
            token = VerificationToken.query.filter_by(
                token=token_string, token_type="password_reset"
            ).first()

            if not token:
                raise ValueError("Invalid reset token")

            if token.is_expired():
                raise ValueError("Reset token has expired")

            if token.is_used:
                raise ValueError("Reset token has already been used")

            # Update user password
            user = User.query.get(token.user_id)
            if not user:
                raise ValueError("User not found")

            user.password_hash = AuthService.hash_password(new_password)
            token.is_used = True
            token.used_at = datetime.utcnow()
            db.session.commit()

            logger.info(f"Password reset for user: {user.email}")

            return {"message": "Password reset successfully"}

        except Exception as e:
            db.session.rollback()
            logger.error(f"Failed to reset password: {str(e)}")
            raise

    @staticmethod
    def change_password(user_id, current_password, new_password):
        """Change user password."""
        try:
            # Find user
            user = User.query.get(user_id)
            if not user:
                raise ValueError("User not found")

            # Verify current password
            if not AuthService.verify_password(current_password, user.password_hash):
                raise ValueError("Current password is incorrect")

            # Update password
            user.password_hash = AuthService.hash_password(new_password)
            db.session.commit()

            logger.info(f"Password changed for user: {user.email}")

            return {"message": "Password changed successfully"}

        except Exception as e:
            db.session.rollback()
            logger.error(f"Failed to change password: {str(e)}")
            raise

    @staticmethod
    def create_verification_token(user_id, token_type, expires_in_minutes):
        """Create verification token."""
        try:
            import uuid

            token_string = str(uuid.uuid4())
            expires_at = datetime.utcnow() + timedelta(minutes=expires_in_minutes)

            token = VerificationToken(
                user_id=user_id,
                token_type=token_type,
                token=token_string,
                expires_at=expires_at,
                is_used=False,
            )
            db.session.add(token)
            db.session.commit()

            logger.debug(f"Verification token created: {token_type}")
            return token

        except Exception as e:
            db.session.rollback()
            logger.error(f"Failed to create verification token: {str(e)}")
            raise

    @staticmethod
    def sanitize_user(user):
        """Remove sensitive fields from user object."""
        if not user:
            return None
        return user.to_dict()
